### PR TITLE
Handle shared params & non-persistent buffer

### PR DESCRIPTION
**Pull request by mavenlin**
`t2j_module` visits each module to set the buffer and params.
There are special cases for both.

1. a parameter could be shared between two module, e.g. `llm_head.weight` sharing that of `embedding.weights`, but it only appears in `state_dict` once.
2. a buffer could be a nonpersistent buffer, which means it won't appear in `state_dict`.

### CODE REVIEW

File: torch2jax/__init__.py [hunk 1]
```diff
@@ -1134,12 +1134,26 @@ def f(*args, state_dict={}, rng=None, return_state_dict=False):
...
 
-      for name, _ in m.named_buffers(recurse=False):
-        m._buffers[name] = Torchish(state_dict[".".join(prefix + [name])])
+    def visit(m, prefix):
+      for name, param in m.named_parameters(recurse=False):
+        if param in reverse_dict:
+          # sometimes parameters are shared,
+          # e.g. llm head shares embedding weights
+          # we won't be able to get "llm.head" in the state_dict in this case.
+          # because the underlying parameter is only registered once as "embedding.weight".
```
Comment (samuela) [inline torch2jax/__init__.py:1145]: i have a few questions about this aspect. please pardon my pytorch ignorance,
- iiuc you're saying that in some cases `named_parameters()` does not line up one-to-one with `state_dict().keys()`. is that correct? in this scenario how does pytorch negotiate what values go where in its implementation of `load_state_dict()`? presumably this is the logic to piggyback off of
- if the underlying parameter is shared, does that mean that gradients should be accumulated/shared as well? or what about in-place operations on the `Parameter`? in the in-place ops case we'll want to reuse the same `Torchish` instance

**Review by samuela — COMMENTED** [review]
Thanks for submitting this @mavenlin ! some interesting aspects of the pytorch Module api that i was not previously aware of. let's add tests and get this merged

File: torch2jax/__init__.py (hunk 1 above)
Comment (mavenlin) [reply torch2jax/__init__.py:1145]: 1. Yes. I think via the key? no persistent buffer does not need loading (computed during init). And for parameters if they are shared, they point to the same object that is only registered once.
2. Yes the gradients are accumulated I assume. I don't know how inplace is now having no problem with the gradient, But I still think the real situation could be more complex and inplace could fail for Jax in some cases.

Comment (samuela) [reply torch2jax/__init__.py:1145]: EDIT: may need to be `torch.nn.Parameter -> Torchish` but you get the idea hopefully

Comment (mavenlin) [reply torch2jax/__init__.py:1145]: Let me clarify a bit after some research.
1. `state_dict()` and `load_state_dict()` will have two copies as if they were different.
2. the weight tying happens during initialization, which makes the same parameter gets registered twice in two modules.
3. `named_parameters` or `parameters` will only retain one copy of them.

**Comment by mavenlin** [comment]
For this, I think it is good to go too.
But the solution is not complete, as the `register_buffer` can be invoked in the forward pass, and currently `Torchish` can not be registered. There's not much we could do before https://github.com/pytorch/pytorch/pull/159455 goes into pytorch.

File: torch2jax/__init__.py (hunk 1 above)
Comment (samuela) [reply torch2jax/__init__.py:1145]: Thanks for digging into this and writing up a clear breakdown of what we're up against. Based on this, I'm happy with the solution that you've come up with here.

> The above said, the input to the jax function should not be constructed from t2j on state_dict, but instead t2j on the named_parameters & named_buffers. Do you think it is a good idea to help the user construct them? i.e. t2j_module returns f, param, state, like haiku's [transform_with_state](https://dm-haiku.readthedocs.io/en/latest/api.html#transform-with-state), where we distinguish differentiable / non-differentiable variables into param and state respectively.

What exactly is the API that you're proposing? I would be curious to explore alternatives, but I'm not immediately seeing the advantage of switching. Using state_dict has the advantage that users can load them directly from file/an existing model without having to manipulate them before handing it off to torch2jax.

**Comment by samuela** [comment]
> But the solution is not complete, as the register_buffer can be invoked in the forward pass, and currently Torchish can not be registered. There's not much we could do before https://github.com/pytorch/pytorch/pull/159455 goes into pytorch.

Ah, good catch. Nice to see that pytorch fixed this

**Comment by samuela** [comment]
> Haha, that's me sending the PR.

I saw... good stuff :)
